substrate/primitives/core/src/paired_crypto.rs

```diff
@@ -0,0 +1,658 @@
+	where
+		D: Deserializer<'de>,
+	{
+		let bytes = array_bytes::hex2bytes(&String::deserialize(deserializer)?)
+			.map_err(|e| de::Error::custom(format!("{:?}", e)))?;
+		Signature::<LEFT_PLUS_RIGHT_LEN>::try_from(bytes.as_ref()).map_err(|e| {
+			de::Error::custom(format!("Error converting deserialized data into signature: {:?}", e))
+		})
+	}
+}
+
+impl<const LEFT_PLUS_RIGHT_LEN: usize> From<Signature<LEFT_PLUS_RIGHT_LEN>>
+	for [u8; LEFT_PLUS_RIGHT_LEN]
+{
+	fn from(signature: Signature<LEFT_PLUS_RIGHT_LEN>) -> [u8; LEFT_PLUS_RIGHT_LEN] {
+		signature.0
+	}
+}
+
+impl<const LEFT_PLUS_RIGHT_LEN: usize> sp_std::fmt::Debug for Signature<LEFT_PLUS_RIGHT_LEN>
+where
+	[u8; LEFT_PLUS_RIGHT_LEN]: crate::hexdisplay::AsBytesRef,
+{
+	#[cfg(feature = "std")]
+	fn fmt(&self, f: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+		write!(f, "{}", crate::hexdisplay::HexDisplay::from(&self.0))
+	}
+
+	#[cfg(not(feature = "std"))]
+	fn fmt(&self, _: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+		Ok(())
+	}
+}
+
+impl<const LEFT_PLUS_RIGHT_LEN: usize> UncheckedFrom<[u8; LEFT_PLUS_RIGHT_LEN]>
+	for Signature<LEFT_PLUS_RIGHT_LEN>
+{
+	fn unchecked_from(data: [u8; LEFT_PLUS_RIGHT_LEN]) -> Self {
+		Signature(data)
+	}
+}
+
+/// A key pair.
+#[cfg(feature = "full_crypto")]
+#[derive(Clone)]
+pub struct Pair<
+	LeftPair: PairT,
+	RightPair: PairT,
+	const PUBLIC_KEY_LEN: usize,
+	const SIGNATURE_LEN: usize,
+> {
+	left: LeftPair,
+	right: RightPair,
+}
+
+#[cfg(feature = "full_crypto")]
+impl<
+		LeftPair: PairT,
+		RightPair: PairT,
+		const PUBLIC_KEY_LEN: usize,
+		const SIGNATURE_LEN: usize,
+	> PairT for Pair<LeftPair, RightPair, PUBLIC_KEY_LEN, SIGNATURE_LEN>
+where
+	Pair<LeftPair, RightPair, PUBLIC_KEY_LEN, SIGNATURE_LEN>: CryptoType,
+	LeftPair::Signature: SignatureBound,
+	RightPair::Signature: SignatureBound,
+	Public<PUBLIC_KEY_LEN>: CryptoType,
+	LeftPair::Seed: From<Seed> + Into<Seed>,
+	RightPair::Seed: From<Seed> + Into<Seed>,
+{
+	type Seed = Seed;
+	type Public = Public<PUBLIC_KEY_LEN>;
+	type Signature = Signature<SIGNATURE_LEN>;
+
+	fn from_seed_slice(seed_slice: &[u8]) -> Result<Self, SecretStringError> {
+		if seed_slice.len() != SECURE_SEED_LEN {
+			return Err(SecretStringError::InvalidSeedLength)
+		}
+		let left = LeftPair::from_seed_slice(&seed_slice)?;
+		let right = RightPair::from_seed_slice(&seed_slice)?;
+		Ok(Pair { left, right })
+	}
+
+	fn derive<Iter: Iterator<Item = DeriveJunction>>(
+		&self,
+		path: Iter,
+		seed: Option<Self::Seed>,
+	) -> Result<(Self, Option<Self::Seed>), DeriveError> {
+		let path: Vec<_> = path.collect();
+
+		let left = self.left.derive(path.iter().cloned(), seed.map(|s| s.into()))?;
+		let right = self.right.derive(path.into_iter(), seed.map(|s| s.into()))?;
+
+		let seed = match (left.1, right.1) {
+			(Some(l), Some(r)) if l.as_ref() == r.as_ref() => Some(l.into()),
+			_ => None,
+		};
+
+		Ok((Self { left: left.0, right: right.0 }, seed))
```

Comment:
Mmm... here if the two crypto schemes have different ways to derive the seed we end up with a `None` seed. 
I don't like this a lot. But if is good for the current application I'm not going to block things for this.

Would be nice to at least write this thing explicitly over this method docs.

Reply:
@davxy something like https://github.com/paritytech/polkadot-sdk/pull/1705#discussion_r1355039462?
can also integrate directly into [`PairT::derive`](https://github.com/w3f/polkadot-sdk/blob/a069dcd810dc5f7dfea2f1fb5c259e3369961039/substrate/primitives/core/src/crypto.rs#L892-L893) documentation, although it's specific to the implementation here.

Reply:
The only way to be able to return a seed is to make the paired_crypto seed 64 bytes which is not inline with all other crypto schemes (and make it not composable i.e. you can't have a pair of paired schemes).

There is a comment in crypto.rs which says in some crypto schemes it is not possible to provide a direct seed after derivation: 

https://github.com/w3f/polkadot-sdk//blob/skalman-paired-crypto-scheme/substrate/primitives/core/src/crypto.rs#L923

We can also add paired_crypto scheme to that comment perhaps.